lib/IRGen/GenMeta.cpp

```diff
@@ -2620,18 +2561,120 @@ namespace {
+      B.addRelativeAddress(function);
+    }
+
+    void addDestructorFunction() {
+      auto function = getAddrOfDestructorFunction(IGM, Target);
+      B.addRelativeAddressOrNull(function ? *function : nullptr);
+    }
+
+    void addIVarDestroyer() {
+      auto function = IGM.getAddrOfIVarInitDestroy(Target,
+                                                   /*isDestroyer=*/ true,
+                                                   /*isForeign=*/ false,
+                                                   NotForDefinition);
+      B.addRelativeAddressOrNull(function ? *function : nullptr);
+    }
+
+    void addClassFlags() {
+      B.addInt32((uint32_t) getClassFlags(Target));
+    }
+
+    void addClassDataPointer() {
+      auto data = (IGM.ObjCInterop
+                   ? emitClassPrivateData(IGM, Target)
+                   : nullptr);
+      B.addRelativeAddressOrNull(data);
+    }
+
+    void addMetaclass() {
+      auto metaclass = (IGM.ObjCInterop
+                        ? IGM.getAddrOfMetaclassObject(Target, NotForDefinition)
+                        : nullptr);
+      B.addRelativeAddressOrNull(metaclass);
+    }
+
+    void createMetadataAccessFunction() {
+      assert(doesClassMetadataRequireRelocation(IGM, Target));
+
+      assert(!Target->isGenericContext());
+      emitClassMetadataBaseOffset(IGM, Target);
+      createNonGenericMetadataAccessFunction(IGM, Target);
+
+      emitMetadataCompletionFunction(
+          IGM, Target,
+          [&](IRGenFunction &IGF, llvm::Value *metadata,
+              MetadataDependencyCollector *collector) {
+        emitInitializeClassMetadata(IGF, Target, FieldLayout, metadata,
+                                    collector);
+      });
+
+      emitRelocationFunction();
+    }
+
+  private:
+    /// Emit the create function for a class with resilient ancestry.
+    void emitRelocationFunction() {
+      // using MetadataRelocator =
+      //   Metadata *(TypeContextDescriptor *type, void *pattern);
+      llvm::Function *f =
+        IGM.getAddrOfTypeMetadataInstantiationFunction(Target, ForDefinition);
+      f->setAttributes(IGM.constructInitialAttributes());
+
+      IRGenFunction IGF(IGM, f);
+
+      // Skip instrumentation when building for TSan to avoid false positives.
+      // The synchronization for this happens in the Runtime and we do not see it.
+      if (IGM.IRGen.Opts.Sanitizers & SanitizerKind::Thread)
+        f->removeFnAttr(llvm::Attribute::SanitizeThread);
+
+      if (IGM.DebugInfo)
+        IGM.DebugInfo->emitArtificialFunction(IGF, f);
+
+      Explosion params = IGF.collectParameters();
+      llvm::Value *descriptor = params.claimNext();
+      llvm::Value *pattern = params.claimNext();
+
+      // Allocate class metadata using the pattern we emitted.
+      llvm::Value *metadata =
+        IGF.Builder.CreateCall(IGF.IGM.getRelocateClassMetadataFn(),
+                               {descriptor, pattern});
+
+      IGF.Builder.CreateRet(metadata);
+    }
```

Comment:
If this body is expected to be this simple in most situations, should we just allow it to be null, with the semantics that the runtime will just use `swift_relocateClassMetadata` by default?

That same question might reasonably be asked of generic types, I suppose; I don't remember what we do in their allocation functions that can't reasonably be defaulted.